### DISCUSSION TITLE
Performance of Ray plasma vs MPI communication

### DISCUSSION THREAD

**Discussion by YarShev** [mpi4py/mpi4py#332] [General]
Hi folks,

I posted a discussion in mpi4py groups about two months ago but haven't got a response there yet. Not sure if you have seen the discussion so decided to also put it here. Would appreciate any feedback.

link - https://groups.google.com/g/mpi4py/c/BjNn3xflMho

Thanks in advance.

**Comment by dalcinl** [comment]
Google Groups is not working well for me. Your email never reached my Inbox (perhaps it went to my spam folder, but if that is the case, after two months is surely gone, and I cannot verify).

Regarding ultimate MPI performance, you should really look for answers from the developers of the backend MPI implementation you are using. Moreover, MPI implementations have different runtime parameters to set that may affect performance. You did not mention what MPI implementation you are using, that's a crucial piece of information. Also, should I assume that all your proceses are running on the same compute node? If that is the case, and if you are using Open MPI, you should make sure you are truly using shared memory for communication (they call it the `vader` MCA component). Open MPI can use CMA (run `ompi_info --param btl vader -l 3`, the `btl_vader_single_copy_mechanism` should be `cma`). CMA can improve performance substantially, as it allows for direct memory copy from one process to another (without using intermediate shared memory, thus called "single-copy"). If you are using MPICH, I don't think they have support for CMA.

All that been said, take into account that if  you are using point-to point MPI communication, even within the same compute node, then MPI will ALWAYS have to copy memory from one process to another. Ray uses a shared-memory object store with readonly objects. I'm not aware of the low level details, but I guess that they copy memory on `put()`, but maybe not on `get()`.  You should read and understand the serialization section in Ray docs.

Out of curiosity, I pip-installed Ray on my workstation and ran your code:
```
$ python testray.py | grep 268435568
2023-03-21 09:36:11,099	INFO worker.py:1553 -- Started a local Ray instance.
time of ray.put for data size 268435568 = 0.05273103713989258
```
So the time for put is about 52 milliseconds.

Running the following in my command line (IIRC, requires mpi4py from git master) with MPICH from Fedora 37 as the MPI backend.
```
$ mpiexec -n 2 python -m mpi4py.bench pingpong -m 268435456 -n 268435456
# MPI PingPong Test
# Size [B]  Bandwidth [MB/s] | Time Mean [s] ± StdDev [s]  Samples
 268435456           9935.04 | 2.7019067e-02 ± 1.1359e-02       10
```
So that's about 27 milliseconds. 

Of course, my test is different than yours, but you have to grant me that your testing is a bit too basic (you do just one iteration with no repetitions).

**Reply by YarShev** [reply]
Hi @dalcinl, thanks for your response!

I had `mpi4py-mpich` from PyPI and yes, all processes were running on the same node.

`ray.put` makes a copy of data in order to store it into the shared object storage, whereas `ray.get` allows to retrieve data without copy (e.g., for NumPy arrays). That is probably the case why Ray communication gets faster than MPI starting from a certain point on my side. I will definitely try Open MPI with enabling the MCA parameters.

Duplicating the discussion here in order for you to try the MPI and Ray examples out.

>  It was found out that MPI is faster than Ray with respect to data transfer for size <= ~1 MB from one process to another. However, when increasing the data size > ~1 MB MPI gets slower than Ray.

**Ray**
<details>

```python
# launch `python script1.py`
import numpy as np
import ray
import time
import sys


ray.init(num_cpus=2)

sizes = [
        1024 * 128,  # ~128 KB
        1024 * 256,  # ~256 KB
        1024**2,  # ~1 MB
        1024**2 * 128,  # ~128 MB
        1024**2 * 256,  # ~256 MB
    ]
type_size = 8 # because numpy array has int64 dtype by default
sizes = [size // type_size for size in sizes]

input_obj_refs = []
for size in sizes:
    arr = np.zeros(size)
    t0 = time.time()
    ref = ray.put(arr) # data copy to be put in Ray Plasma object store
    t1 = time.time()
    print(f"time of ray.put for data size {sys.getsizeof(arr)} =", t1 - t0)
    input_obj_refs.append(ref)

@ray.remote
def foo(o_refs):
    t0 = time.time()
    arr = ray.get(o_refs[0]) # data retrieving with zero-copy
    t1 = time.time()
    print(f"time of ray.get for data size {sys.getsizeof(arr)} =", t1 - t0)
    return arr

output_obj_refs = [foo.remote([ref]) for ref in input_obj_refs]

# output
time of ray.put for data size 131184 = 0.0011148452758789062
time of ray.put for data size 262256 = 0.0005316734313964844
time of ray.put for data size 1048688 = 0.0011167526245117188
time of ray.put for data size 134217840 = 0.02989053726196289
time of ray.put for data size 268435568 = 0.0584867000579834
time of ray.get for data size 112 = 0.0009331703186035156
time of ray.get for data size 112 = 0.00015425682067871094
time of ray.get for data size 112 = 0.0001499652862548828
time of ray.get for data size 112 = 0.0009522438049316406
time of ray.get for data size 112 = 0.00015735626220703125
```

</details>

> As can be seen from the output above, time of ray.put depends on the data size as Ray creates a copy of data to be put into plasma. Time of ray.get is more or less the same as Ray retrieves the data from the shared object storage with zero copy.

**MPI**
<details>

```python
# launch `mpiexec -n 2 python script2.py`
import time
import numpy as np
import sys
from mpi4py import MPI

comm = MPI.COMM_WORLD
rank = comm.Get_rank()

sizes = [
    1024 * 128,  # ~128 KB
    1024 * 256,  # ~256 KB
    1024**2,  # ~1 MB
    1024**2 * 128,  # ~128 MB
    1024**2 * 256,  # ~256 MB
]
type_size = 8 # because numpy array has int64 dtype by default
sizes = [size // type_size for size in sizes]
arrays = [np.zeros(size) for size in sizes]

if rank == 0:
    for i in range(len(arrays)):
        t0 = time.time()
        comm.Send(arrays[i], dest=1)
        t1 = time.time()
        print(f"Send time for data size {sys.getsizeof(arrays[i])} =", t1 - t0)
elif rank == 1:
    for i in range(len(arrays)):
        t0 = time.time()
        comm.Recv(arrays[i], source=0)
        t1 = time.time()
        print(f"Recv time for data size {sys.getsizeof(arrays[i])} =", t1 - t0)

# output
Send time for data size 131184 = 0.00021719932556152344
Send time for data size 262256 = 0.00012946128845214844
Send time for data size 1048688 = 0.00046062469482421875
Send time for data size 134217840 = 0.07068681716918945
Send time for data size 268435568 = 0.14120078086853027
Recv time for data size 131184 = 0.00021886825561523438
Recv time for data size 262256 = 0.00014495849609375
Recv time for data size 1048688 = 0.0005681514739990234
Recv time for data size 134217840 = 0.07068228721618652
Recv time for data size 268435568 = 0.14120888710021973
```

</details>

> As can be seen from the output above, MPI gets slower than Ray when data sizes get increased.

Do you think if it is worth using MPI-3 Shared Memory to get MPI faster than Ray?

**Reply by dalcinl** [reply]
> I had `mpi4py-mpich` from PyPI and yes,

That package is not under the control of the mpi4py project. I take absolutely no responsibility for its contents.

> Do you think if it is worth using MPI-3 Shared Memory to get MPI faster than Ray?

Well, I'm not really an expert on MPI shared memory, we should get @jeffhammond involved. That being said, I believe I've learned enough from Jeff as to provide you some suggestions for things to try.

* First, you use `Win.Allocate_shared(size=n, comm=comm)` to allocate a shared memory block of `n` bytes locally. Each rank in `comm` will allocate its own block, `n` can be different across ranks. The return is a ~buffer-like~ window object, that is also a buffer-like object.
* Second, you can use `Win.Shared_query()` to query remote memory blocks. Again, you get a buffer-like object.
* Afterwards, you can use the buffer-like objects to create NumPy arrays using [`numpy.frombuffer`](https://numpy.org/doc/stable/reference/generated/numpy.frombuffer.html).
* Finally, you can use direct loads and stores using these NumPy arrays.

This approach should be as fast as `memcpy()` can get. I'll try to put an example here for you to try.

**Reply by dalcinl** [reply]
Here you have.

```python
import numpy as np
from mpi4py import MPI

dtype = np.dtype(np.float64)
itemsize = dtype.itemsize
length = 1024**2 * 256 // itemsize

comm = MPI.COMM_WORLD
assert comm.size == 2

win = MPI.Win.Allocate_shared(length * itemsize, itemsize, comm=comm)
blocks = [win.Shared_query(i) for i in range(comm.size)]
arrays = [np.frombuffer(buf, dtype=dtype) for buf, _ in blocks]

loc = arrays[comm.rank]
rem = arrays[comm.rank-1]

loc.fill(-1)
rem.fill(42)
comm.Barrier()
comm.Barrier()

t1 = MPI.Wtime()
loc[:] = rem
t2 = MPI.Wtime()
print(comm.rank, loc.nbytes, (t2-t1))
```
I would expect you to get about the same performance as with Ray.

**Reply by YarShev** [reply]
@dalcinl, thanks for providing the example! I will try it out. I started with Open MPI exploration. How can I enable CMA for it? This is the output of `ompi_info --param btl vader -l 3` I get.

```bash
$ ompi_info --param btl vader -l 3
                 MCA btl: vader (MCA v2.1.0, API v3.1.0, Component v4.1.5)
           MCA btl vader: ---------------------------------------------------
           MCA btl vader: parameter "btl_vader_single_copy_mechanism"
                          (current value: "emulated", data source: default,
                          level: 3 user/all, type: int)
                          Single copy mechanism to use (defaults to best
                          available)
                          Valid values: 4:"emulated", 3:"none"
           MCA btl vader: parameter "btl_vader_backing_directory" (current
                          value: "/dev/shm", data source: default, level: 3
                          user/all, type: string)
                          Directory to place backing files for shared memory
                          communication. This directory should be on a local
                          filesystem such as /tmp or /dev/shm (default:
                          (linux) /dev/shm, (others) session directory)
```

**Comment by dalcinl** [comment]
Mmm... that does not look good. There is no `"cma"` value listed.
Where did you get your Open MPI binaries from?

Look at my output (Open MPI from Fedora 37 system rpm packages)
```
$ ompi_info --param btl vader -l 3
                 MCA btl: vader (MCA v2.1.0, API v3.1.0, Component v4.1.4)
           MCA btl vader: ---------------------------------------------------
           MCA btl vader: parameter "btl_vader_single_copy_mechanism"
                          (current value: "cma", data source: default, level:
                          3 user/all, type: int)
                          Single copy mechanism to use (defaults to best
                          available)
                          Valid values: 1:"cma", 4:"emulated", 3:"none"
```
Note `current value: "cma"` in the output

**Reply by YarShev** [reply]
I installed it with this way.

```python
mamba install -c conda-forge mpi4py openmpi
```

**Reply by YarShev** [reply]
Is this a correct way to install OpenMPI?

**Reply by dalcinl** [reply]
What you did is correct. But looks like Open MPI as built in conda-forge does not have support for cross-memory attach. 

Maybe the reason is that the Linux kernel in CI build runners is too old, or somehow it does not expose or block CMA features to running applications. Therefore, Open MPI's configure cannot find the required pieces to build CMA support in vader. 

@jsquyres @devreal Could you guys please help us here? Any tips?

**Reply by jsquyres** [reply]
I'm afraid I don't know much (read: anything) about the conda method for installing Open MPI.

When Open MPI is built, it looks for the stuff it needs to build CMA support, namely: 

* The `sys/prctl.h` header file
* The function `process_vm_readv`
* syscall ID's for Linux CMA APIs

If all of these things can't be found, Open MPI will skip building CMA support.  Note that CMA support is only available on modern flavors of Linux (e.g., it's not available on MacOS).  As Lisandro said, if conda is simply shipping you Open MPI binaries, perhaps the Linux machines they built Open MPI on were too old / didn't have both things needed for CMA support.  If conda is building Open MPI from source, you should be able to look the 78 miles of output from the build and look closer in the output from Open MPI's configure script.  You should be able to see something like this:

```
--- MCA component btl:vader (m4 configuration macro)
checking for MCA component btl:vader compile mode... dso
checking for Cray XPMEM support... checking for CRAY_XPMEM... no
checking --with-xpmem value... simple ok (unspecified value)
checking --with-xpmem-libdir value... simple ok (unspecified value)
checking xpmem.h usability... no
checking xpmem.h presence... no
checking for xpmem.h... no
checking --with-knem value... simple ok (unspecified value)
checking knem_io.h usability... no
checking knem_io.h presence... no
checking for knem_io.h... no
checking sys/prctl.h usability... yes
checking sys/prctl.h presence... yes
checking for sys/prctl.h... yes
checking for process_vm_readv... no
checking if internal syscall numbers for Linux CMA work... yes
checking for sys/prctl.h... (cached) yes
checking if MCA component btl:vader can compile... yes
```

Notice the checks in there for `sys/prctl.h` and `process_vm_ready` and the syscall numbers -- they all say "yes".  If any of them say "no", then Open MPI will simply skip building CMA support.

If conda-installed Open MPI isn't giving you what you want, you could install Open MPI from source yourself, and be in finer-grained control of what Open MPI installation.  A good place to start would be to find the `configure` line from the conda build and simply run that in an Open MPI tarball download (the most recent stable release is Open MPI v4.1.5).  Then check the `configure` output for the CMA stuff I listed above to ensure it will build CMA support for you.

**Reply by dalcinl** [reply]
> If conda is building Open MPI from source,

That is indeed the case. Now we know exactly how where to [look](https://dev.azure.com/conda-forge/84710dde-1620-425b-80d0-4cf5baca359d/_apis/build/builds/675620/logs/43). Thanks for the help, @jsquyres!

This is the relevant output:
```
2023-03-18T13:14:36.0085291Z --- MCA component btl:vader (m4 configuration macro)
2023-03-18T13:14:36.0095001Z checking for MCA component btl:vader compile mode... dso
2023-03-18T13:14:36.0253494Z checking for Cray XPMEM support... checking for CRAY_XPMEM... no
2023-03-18T13:14:36.0293522Z checking --with-xpmem value... simple ok (unspecified value)
2023-03-18T13:14:36.0301385Z checking --with-xpmem-libdir value... simple ok (unspecified value)
2023-03-18T13:14:36.0885235Z checking for xpmem.h... no
2023-03-18T13:14:36.1056294Z checking --with-knem value... simple ok (unspecified value)
2023-03-18T13:14:36.1662700Z checking for knem_io.h... no
2023-03-18T13:14:36.2372061Z checking for sys/prctl.h... yes
2023-03-18T13:14:36.2967250Z checking for process_vm_readv... no
2023-03-18T13:14:36.3936339Z checking if internal syscall numbers for Linux CMA work... no
2023-03-18T13:14:36.4292522Z checking if MCA component btl:vader can compile... yes
```
As you can see, there is a line with `checking for process_vm_readv... no`.  Maybe this happens because the build is running within a Docker container. @YarShev If you want this to be investigated further, you should submit an issue to [conda-forge/openmp-feedstock](https://github.com/conda-forge/openmpi-feedstock). I don't have enough expertise about the low level details of conda-forge infrastructure to help you further.

PS: @jsquyres I believe that `process_vm_read`  should be available by including `sys/uio.h` (see [manpage](https://man7.org/linux/man-pages/man2/process_vm_readv.2.html)). Anyway, maybe your configure checks do really depend on including the proper sys header.

**Reply by YarShev** [reply]
@dalcinl, @jsquyres, thanks for the input. I created an issue in openmpi-feedstock. Hope it will be fixed there.

https://github.com/conda-forge/openmpi-feedstock/issues/118

**Reply by YarShev** [reply]
> CMA can improve performance substantially, as it allows for direct memory copy from one process to another (without using intermediate shared memory, thus called "single-copy"). If you are using MPICH, I don't think they have support for CMA.

@dalcinl, btw, I would like to ask you about the statement above. What is the benefit of using intermediate shared memory? and if some MPI implementations do not have support for CMA, why they haven't added the support for it yet?

**Reply by dalcinl** [reply]
There is absolutely no benefit of using intermediate shared memory. Before CMA availability in the Linux kernel (or similar mechanisms like XPMEM), the only way two processes had copy local memory each other is to use intermediate shared memory blocks. The sender copies the send buffer to shmem, and the receiver copies from shmem to the recv buffer, thus you a two-copy mechanism. With CMA, a process can directly copy from/to another cooperating process.

Regarding why other MPI implementations like MPICH have not added support yet for CAM, well, you should really ask them. @hzhou Is adding cross-memory attach support in MPICH's radar?

I can speculate, though. MPI primarily cares about large scale and off-node communication, and ultimate intra-node communication performance is not that important.  User needs are infinite, development teams have limited resources, and MPI devs have to prioritize. If this is really a big deal to business customers, then maybe you should contribute back to the open-source projects, either financially or with working quality code.

**Reply by jeffhammond** [reply]
UCX supports CMA (https://github.com/openucx/ucx/blob/master/docs/source/faq.md). Therefore, both MPICH and OMPI will use CMA via UCX assuming everything is configured appropriately.

**Reply by dalcinl** [reply]
> UCX supports CMA

Of course! I forgot about that! Thanks for the reminder. Unfortunately, the conda-forge MPICH package is still stuck on `--with-device=ch3:nemesis`.

@YarShev You can still build MPICH (or Open MPI) yourself, and install all the other conda-forge packages to use that external MPICH (or Open MPI) build. Just read the [docs](https://conda-forge.org/docs/user/tipsandtricks.html#using-external-message-passing-interface-mpi-libraries).
